pkg/cache/stats_counter.go

```diff
@@ -0,0 +1,216 @@
+
+func WithTrackLoadingCacheStats() Option {
+	return func(c *burrowStatsCounter) {
+		c.trackLoadingCacheStats = true
+		c.loadSuccess = promauto.With(c.reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_load_success_total",
+			Help: "Total number of successful cache loads.",
+		})
+		c.loadError = promauto.With(c.reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_load_error_total",
+			Help: "Total number of cache load errors.",
+		})
+		c.loadTotalTime = promauto.With(c.reg).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cache_load_duration_seconds",
+			Help:    "Total time spent loading cache.",
+			Buckets: []float64{0.001, 0.002, 0.005, 0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1},
+		})
+	}
+}
+
+// NewBurrowStatsCounter creates a new StatsCounter.
+// RecordLoadSuccess and RecordLoadError methods are called by Get methods on a successful and failed load respectively.
+// Get method only called by LoadingCache implementation.
+func NewBurrowStatsCounter(logger log.Logger, reg prometheus.Registerer, name string, options ...Option) *burrowStatsCounter {
+	reg = prometheus.WrapRegistererWith(prometheus.Labels{"cache": name}, reg)
+	s := &burrowStatsCounter{
+		logger: logger,
+		reg:    reg,
+
+		hits: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_hits_total",
+			Help: "Total number of cache hits.",
+		}),
+		miss: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_miss_total",
+			Help: "Total number of cache misses.",
+		}),
+		eviction: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_evictions_total",
+			Help: "Total number of cache evictions.",
+		}),
+	}
+	for _, opt := range options {
+		opt(s)
+	}
+	return s
+}
+
+// RecordHits records the number of hits.
+// This method is called by Get and GetIfPresent methods on a cache hit.
+func (c *burrowStatsCounter) RecordHits(hits uint64) {
+	c.hits.Add(float64(hits))
+}
+
+// RecordMisses records the number of misses.
+// This method is called by Get and GetIfPresent methods method on a cache miss.
+func (c *burrowStatsCounter) RecordMisses(miss uint64) {
+	c.miss.Add(float64(miss))
+}
+
+// RecordLoadSuccess records the number of successful loads.
+// This method is called by Get methods on a successful load.
+func (c *burrowStatsCounter) RecordLoadSuccess(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadSuccess.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordLoadError records the number of failed loads.
+// This method is called by Get methods on a failed load.
+func (c *burrowStatsCounter) RecordLoadError(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadError.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordEviction records the number of evictions.
+func (c *burrowStatsCounter) RecordEviction() {
+	c.eviction.Inc()
+}
+
+// Snapshot records the current stats.
+// This method is called only by Stats method.
```

Comment:
Do I understand correctly that this implements some interface? If so please add it to the code comment, I read this probably 5 times and didn't understand why this is here.

Reply:
Sorry about not being clear enough. I will update the PR description and the code comments. 
Yes, this is for conforming an interface that's required by the cache and this particular function will only be used while debugging or it's explicitly called for another reason. Which we don't have in our codebase and the cache package also doesn't call it. I wanted to denote it because it's rather a costly method.